pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,315 @@
+use core::ops::{AddAssign, BitOrAssign, ShlAssign};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_arithmetic::ArithmeticError;
+use substrate_fixed::{
+	traits::{Fixed, FixedSigned, ToFixed},
+	transcendental::{exp, ln, sqrt},
+	types::I9F23,
+};
+
+/// An enumeration representing different types of bonding curves.
+///
+/// This enum is generic over the type `F`, which represents the type of the coefficients used in the bonding functions.
+///
+/// # Variants
+/// - `PolynomialFunction`: Represents a polynomial bonding function with parameters of type `PolynomialFunctionParameters<F>`.
+/// - `SquareRootBondingFunction`: Represents a square root bonding function with parameters of type `SquareRootFunctionParameters<F>`.
+///
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum Curve<F> {
+	PolynomialFunction(PolynomialFunctionParameters<F>),
+	SquareRootBondingFunction(SquareRootFunctionParameters<F>),
+	LSMR(LSMRFunctionParameters<F>),
+}
+
+/// An enumeration representing the type of operation on the bonding curve.
+///
+/// # Variants
+/// - `Mint`: Represents a minting operation, where new tokens are created.
+/// - `Burn`: Represents a burning operation, where existing tokens are destroyed.
+pub enum DiffKind {
+	Mint,
+	Burn,
+}
+
+impl<F> Curve<F>
+where
+	F: FixedSigned + PartialOrd<I9F23> + From<I9F23>,
```

Comment:
TODO: check if this can be more generic in the different implementation. Only LSMR needs signed integers.